### ISSUE TITLE
Buck not re-compiling on file change

### ISSUE THREAD

**Issue by mread** [facebook/buck#759]
I'm not using buckd.

I have a simple module with a java_test rule. I run `buck test` and the tests pass. I introduce an Assert.fail() in one of the tests and the test fails. I then remove the Assert.fail() but the tests continue to report failure.

A few things are noticeable when the problem occurs:
1. The console output correctly highlights that the module has changed here - `1 UPDATED`

`[-] BUILDING...FINISHED 0.7s [100%] (198/198 JOBS, 1 UPDATED, 0.0% CACHE MISS)`
1. The javac step doesn't get triggered - there's a noticeable pause when the tests are compiled but I don't see it when this problem occurs.
2. The output from the --build-report switch says `FETCHED_FROM_CACHE` even though 1 above says a rule was updated.

```
{
  "success" : true,
  "results" : {
    "//accepted-data-service:test" : {
      "success" : true,
      "type" : "FETCHED_FROM_CACHE",
      "output" : "buck-out/gen/accepted-data-service/lib__test__output/test.jar"
    }
  },
  "failures" : { }
}
```
1. If I `rm -rf buck-out/cache` before each run then the problem goes away. This seems like the smokiest gun.

Any ideas?

Happy to provide more info if I can help.

`buck-out/cache` is a relatively recent introduction (since I was last using buck) - does it cache old artifacts? If we get a cache-hit is it possible that it's not invalidating the downstream steps correctly?

**Comment by Coneko** [comment]
Can you try `--no-results-cache`? Test results have a separate caching mechanism that is much less reliable than the one used for builds, because tests are not assumed to be deterministic.

If that works, this would be a bug in that logic.

**Comment by mread** [comment]
Yeah, I'd tried that. It's not that the test results are cached - they do get re-run. It's just that they're not being recompiled first.

I'm guessing that buck finds a copy of the previously built artifact in the directory cache and decides to use it, hence the `FETCHED_FROM_CACHE`. But then it doesn't use it to run the tests against - not sure how goes astray.

**Comment by mread** [comment]
Additionally, if I turn off directory caching with the following in `.buckconfig` then the problem goes away:
```
[cache]
    mode =
```

**Comment by mread** [comment]
Appears to be fixed by the same change as #827.
